src/test/scala/is/hail/expr/ir/StringSliceSuite.scala

```diff
@@ -0,0 +1,137 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+import is.hail.TestUtils._
+import is.hail.expr.types._
+import org.scalatest.testng.TestNGSuite
+import org.testng.annotations.Test
+
+class StringSliceSuite extends TestNGSuite {
+  @Test def zeroToLengthIsIdentity() {
+    assertEvalsTo(StringSlice(Str("abc"), I32(0), I32(3)), "abc")
+  }
+
+  @Test def simpleSlicesMatchIntuition() {
+    assertEvalsTo(StringSlice(Str("abc"), I32(3), I32(3)), "")
+    assertEvalsTo(StringSlice(Str("abc"), I32(1), I32(3)), "bc")
+    assertEvalsTo(StringSlice(Str("abc"), I32(2), I32(3)), "c")
+    assertEvalsTo(StringSlice(Str("abc"), I32(0), I32(2)), "ab")
+  }
+
+  @Test def sizeZeroSliceIsEmptyString() {
+    assertEvalsTo(StringSlice(Str("abc"), I32(2), I32(2)), "")
+    assertEvalsTo(StringSlice(Str("abc"), I32(1), I32(1)), "")
+    assertEvalsTo(StringSlice(Str("abc"), I32(0), I32(0)), "")
+  }
+
+  @Test def sliceMatchesJavaStringSubstring() {
+    assertEvalsTo(
+      StringSlice(Str("abc"), I32(0), I32(2)),
+      "abc".substring(0, 2))
+    assertEvalsTo(
+      StringSlice(Str("foobarbaz"), I32(3), I32(5)),
+      "foobarbaz".substring(3, 5))
+  }
+
+  @Test def isStrict() {
+    assertEvalsTo(StringSlice(NA(TString()), I32(0), I32(2)), null)
+    assertEvalsTo(StringSlice(NA(TString()), I32(-5), I32(-10)), null)
+  }
+
+  @Test def stringIndexMatchesIntuition() {
+    assertFatal(invoke("[]", Str("abc"), I32(-5)),
+      "string slice out of bounds or invalid: \"abc\"\\[-2:-1\\]")
```

Comment:
I find these error messages confusing since they're translated from what the user entered `[-5]`.

Reply:
I removed string ref. I'll make a separate PR that handles it with better error messages